viewer/src/repository/cad/CachedRepository.ts

```diff
@@ -3,55 +3,105 @@
  */
 
 import { Repository } from './Repository';
-import { MemoryRequestCache } from '../../cache/MemoryRequestCache';
-import { CadModel } from '../../models/cad/CadModel';
-import { Sector, SectorQuads } from '../../models/cad/types';
 import { WantedSector } from '../../data/model/WantedSector';
-import { ParsedSector } from '../../data/model/ParsedSector';
 import { LevelOfDetail } from '../../data/model/LevelOfDetail';
+import { OperatorFunction, pipe, Observable, from, merge } from 'rxjs';
+import { publish, filter, flatMap, map, share, tap, shareReplay, take } from 'rxjs/operators';
+import { ModelDataRetriever } from '../../datasources/ModelDataRetriever';
+import { CADSectorParser } from '../../data/parser/CADSectorParser';
+import { SimpleAndDetailedToSector3D } from '../../data/transformer/three/SimpleAndDetailedToSector3D';
+import { ConsumedSector } from '../../data/model/ConsumedSector';
 
 export class CachedRepository implements Repository {
-  private readonly _detailedCache: MemoryRequestCache<number, null, Promise<Sector>>;
-  private readonly _simpleCache: MemoryRequestCache<number, null, Promise<SectorQuads>>;
-
-  constructor(model: CadModel) {
-    const getDetailedBasic = async (sectorId: number) => {
-      const data = await model.fetchSectorDetailed(sectorId);
-      return model.parseDetailed(sectorId, data);
-    };
-
-    const getSimpleBasic = async (sectorId: number) => {
-      const data = await model.fetchSectorSimple(sectorId);
-      return model.parseSimple(sectorId, data);
-    };
-
-    this._detailedCache = new MemoryRequestCache(getDetailedBasic);
-    this._simpleCache = new MemoryRequestCache(getSimpleBasic);
+  // TODO: j-bjorne 15/04/2020 - Should look into having a smarted cache (current one doesn't clear), should look at MemoryCache implementation
+  private readonly _modelDataCache: Map<string, Observable<ConsumedSector>> = new Map();
+  private readonly _modelDataParser: CADSectorParser;
+  private readonly _modelDataRetriever: ModelDataRetriever;
+  private readonly _modelDataTransformer: SimpleAndDetailedToSector3D;
+
+  constructor(
+    modelDataRetriever: ModelDataRetriever,
+    modelDataParser: CADSectorParser,
+    modelDataTransformer: SimpleAndDetailedToSector3D
+  ) {
+    this._modelDataRetriever = modelDataRetriever;
+    this._modelDataParser = modelDataParser;
+    this._modelDataTransformer = modelDataTransformer;
+  }
+
+  loadSector(): OperatorFunction<WantedSector, ConsumedSector> {
+    return pipe(
+      publish((wantedSectorObservable: Observable<WantedSector>) => {
+        const simpleAndDetailedObservable = wantedSectorObservable.pipe(
+          filter(wantedSector => wantedSector.levelOfDetail !== LevelOfDetail.Discarded),
+          share()
+        );
+
+        const cachedObservable = simpleAndDetailedObservable.pipe(
+          filter(wantedSector => this._modelDataCache.has('' + wantedSector.id + '.' + wantedSector.levelOfDetail)),
```

Comment:
Create `createCacheKey` or something. But can't `WantedSector` be the key?

Reply:
Done